src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,159 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (first == null) {
+            first = new Node<>(null, value, null);
+            size++;
+        } else if (last == null) {
+            last = new Node<>(first, value, null);
+            first.next = last;
+            size++;
+        } else {
+            Node<T> current = last;
+            current.prev.next = current;
+            last = new Node<>(current, value, null);
+            current.next = last;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            Node<T> current = first;
+            current.next.prev = current;
+            first = new Node<>(null, value, current);
+            current.prev = first;
+            size++;
+        } else {
+            Node<T> nodeAfterCurrent = first;
+            nodeAfterCurrent.next.prev = nodeAfterCurrent;
+            nodeAfterCurrent = findNode(nodeAfterCurrent, index);
+            Node<T> nodeBefore = nodeAfterCurrent.prev;
+            Node<T> current = new Node<>(nodeAfterCurrent.prev, value, nodeAfterCurrent);
+            nodeAfterCurrent.prev = current;
+            nodeBefore.next = current;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = first;
+        currentNode.next.prev = currentNode;
+        currentNode = findNode(currentNode, index);
+        return currentNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> nodeSet = first;
+        nodeSet.next.prev = nodeSet;
```

Comment:
Why do we need this?

Reply:
I have used this to create separate reference for nodes (with soo called "iterator" _.next) -  to avoid changes in LinkedList which can make negative effects - JUST MY THOUGHTS